eclipse-jdt/eclipse.jdt.ui
# Add styling enhancements to signatures (JavaElementLinks) in Javadoc. Fixes #1073

## What it does
Adds various enhancements to styling / formatting of signatures in Javadoc presentation (HTML-based) views:
- togglable text wrapping (method parameters, generics type parameters used in method)
- togglable text formatting (generics type parameters used, method parameter names, method name, method class)
- togglable generic type parameters references text coloring
- togglable generic type parameters nesting levels text coloring
- togglable enhancement states: OFF / always ON / ON only on mouse hover over new styling button
- toggable enhancements state preferences saved separately for Javadoc hover view and Javadoc view, color preferences are shared
- styling button also acts as a menu where all aspects of enhancements can be configured : togglable states, color preferences
- defines 4 default colors for type parameters references / levels coloring, further references / levels re-use them repratedly by default
- supports Dark mode (theme) by defining separate default colors and gracefully handling switching themes

Examples:
Default enhancements:
![javadoc_new](https://github.com/eclipse-jdt/eclipse.jdt.ui/assets/36334098/d83620f4-cefd-411e-95d7-348748273b41)
When hovered on styling button
![javadoc_new_colored](https://github.com/eclipse-jdt/eclipse.jdt.ui/assets/36334098/099892a6-8fa6-46c4-97cc-9264be39ae92)


Defaults for colors:
![lightMode-colors-withFormatting](https://github.com/eclipse-jdt/eclipse.jdt.ui/assets/36334098/fd7056c7-d455-4102-91b3-5c61b50645a7)
![darkMode-colors-withFormatting](https://github.com/eclipse-jdt/eclipse.jdt.ui/assets/36334098/caa77c06-3896-4ee3-bfea-7da45734ebe1)

Changing settings example:
![preview_overlay](https://github.com/eclipse-jdt/eclipse.jdt.ui/assets/36334098/90059077-c77b-4f30-b48d-68da6022967e)

![colorPreferencesSubmenu-6colors](https://github.com/eclipse-jdt/eclipse.jdt.ui/assets/36334098/c4d4a02c-b5cf-4bbf-a20a-d41bd59384f0)
![colorPreferencesSubmenu-0colors](https://github.com/eclipse-jdt/eclipse.jdt.ui/assets/36334098/1c6db845-fbbe-4fe9-a6e4-33332bc45f71)




See [Eclipse JDT forums topic](https://www.eclipse.org/forums/index.php?t=msg&th=1114002&goto=1862222&#msg_1862222) I created to communicate goals, approach, decisions taken and progress.

## How to test
At this stage of WIP prototype:
Have Javadoc for various Java source code elements displayed in both Javadoc hover (viewer) and Javadoc view. Interact with new styling menu button (added to toolbars) to form assumptions about what new visual items do & see if things then behave as one expects them to behave. 

## Author checklist


Changes are currently WIP and thus at this stage not meant for thorough code-review, but rather for other people to be able to try them and provide feedback & future code-reviewer to have a look at how changes are  implemented.
